walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,144 @@
+
+    To handle the optional attendees case we essentially run the algorithm
+    twice, thus giving us the same overall complexity and general structure.
+
+    Complexity:
+    Such an approach will take O(nlogn) time to sort and O(n) time to iterate
+    where n is the number of meetings and m is the number of people in the
+    new meeting. The removal will take O(mn) time giving us a complexity of O(mn
+    + nlogn).
+    */
+
+    // Store the times when we cannot meet, essentially take the inverse of this
+    // to determine the times we can meet.
+    Collection<String> meetingAttendees = request.getAttendees();
+    Collection<String> optionalAttendees = request.getOptionalAttendees();
+    Collection<TimeRange> toReturn = new ArrayList<>();
+    // Change to ArrayList so we can sort.
+    ArrayList<Event> eventList = new ArrayList<>(events);
+    // Remove an event if none of the members in the meeting attended it.
+    eventList.removeIf(event
+                       -> !event.getAttendees().stream().anyMatch(
+                           attendee
+                           -> meetingAttendees.contains(attendee) ||
+                                  optionalAttendees.contains(attendee)));
+    Collections.sort(eventList, ORDER_END.reversed());
+    ArrayList<Event> noOptional = new ArrayList<>(eventList);
+    // Remove an event if it is only attended by optional guests.
+    noOptional.removeIf(event
+                        -> event.getAttendees().stream().allMatch(
+                            attendee -> optionalAttendees.contains(attendee)));
+    // Change this to get overlapping events.
+    ArrayList<TimeRange> timesWithOptional = getUnavailableTimes(eventList);
+    ArrayList<TimeRange> timesWithoutOptional = getUnavailableTimes(noOptional);
+    toReturn = getInverse(timesWithOptional, request);
+    if (toReturn.isEmpty() && !timesWithoutOptional.isEmpty()) {
+      toReturn = getInverse(timesWithoutOptional, request);
+    }
+
+    return toReturn;
+  }
+
+  private int getMinStartTime(int currentStartTime, ListIterator<Event> it) {
+    int minStartTime = currentStartTime;
+    while (it.hasNext()) {
+      Event nextEvent = it.next();
+      TimeRange nextEventTime = nextEvent.getWhen();
+      // If an event is in the current range.
+      if (nextEventTime.end() >= currentStartTime) {
+        minStartTime = Math.min(minStartTime, nextEventTime.start());
+      } else {
+        // In this case it is not possible to expand this event any farther,
+        // the iterator already points to the next element we should continue
+        // from.
+        // Reset the iterator so the last tried value isn't skipped.
+        it.previous();
+        break;
+      }
+    }
+    return minStartTime;
+  }
+  // This function assumes that the arguments it is given are in decreasing
+  // order. It thus parses them from end to beginning.
+  private ArrayList<TimeRange> getInverse(ArrayList<TimeRange> fullTimes,
```

Comment:
The function name should be descriptive enough, maybe use getAvailiableTimeSlots(ArrayList<TimeRange> existedMeetingTimes, long requested_meeting_duration), etc, depend on you.
You can order the input meeting arraylist inside the funciton, so user doesn't need to process the input data before call it.